File: 629.cpp
class Solution {
public:
    int MOD = 1e9 + 7;
    
    int kInversePairs(int n, int k) {
        vector<vector<int>> dp(n+1, vector<int>(k+1, -1));
        return helper(dp, n, k);
    }
    int helper(vector<vector<int>>&dp, int n, int k){
        if(k <= 0)
            return k == 0 ? 1 : 0;
        if(dp[n][k] != -1)
            return dp[n][k];
        
        int cnt = 0;
        for(int i = 0; i < n; ++i){
            cnt += (helper(dp, n - 1, k - i) % MOD);
        }
        
        return dp[n][k] = cnt;
    }
};

File: 629.py
class Solution:
    def kInversePairs(self, n: int, k: int) -> int:
        MOD = 10 ** 9 + 7
        @cache
        def helper(nn, kk):
            if kk <= 0:
                return 1
            
            cnt = 0
            for i in range(nn):
                cnt += (helper(nn - i, kk - i) % MOD)
            
            return cnt
        
        return helper(n, k)
                